### PR TITLE
ddl: add slow operations log to add indices, make trouble shooting easier.

### PR DESCRIPTION
**Pull request by winkyao**
## What have you changed? (mandatory)

Log the slow operations in DDL adding indices. It is useful when the adding indices become slow, it can help us to find the problem.

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -259,9 +262,10 @@ var defaultConf = Config{
 			LogRotate: true,
 			MaxSize:   logutil.DefaultLogMaxSize,
 		},
-		SlowThreshold:      300,
-		ExpensiveThreshold: 10000,
-		QueryLogMaxLen:     2048,
+		SlowThreshold:       300,
+		ExpensiveThreshold:  10000,
+		QueryLogMaxLen:      2048,
+		DDLSlowOprThreshold: 300,
```
Comment (shenli) [inline config/config.go]: This is 300ms?

File: config/config.toml.example [hunk 2]
```diff
@@ -63,6 +63,8 @@ expensive-threshold = 10000
 # Maximum query length recorded in log.
 query-log-max-len = 2048
 
+ddl-slow-opr-threshold = 300
```
Comment (shenli) [inline config/config.toml.example]: Add time unit in the comment.

File: ddl/index.go [hunk 3]
```diff
@@ -552,8 +553,13 @@ func (w *addIndexWorker) fetchRowColVals(txn kv.Transaction, taskRange reorgInde
 	w.idxRecords = w.idxRecords[:0]
 	startTime := time.Now()
 	handleOutOfRange := false
+	oprStartTime := time.Now()
 	err := iterateSnapshotRows(w.sessCtx.GetStore(), w.table, txn.StartTS(), taskRange.startHandle,
 		func(handle int64, recordKey kv.Key, rawRow []byte) (bool, error) {
+			oprEndTime := time.Now()
+			w.logSlowOperations(oprEndTime.Sub(oprStartTime), "iterateSnapshotRows in fetchRowColVals is slow", 0)
```
Comment (shenli) [inline ddl/index.go]: "is slow" is redundant with "SLOW-OPERATIONS".

File: ddl/index.go [hunk 4]
```diff
@@ -624,6 +642,7 @@ func (w *addIndexWorker) backfillIndexInTxn(handleRange reorgIndexTask) (nextHan
 		}
 		return nil
 	})
+	w.logSlowOperations(time.Since(oprStartTime), "backfillIndexInTxn is slow", 3000)
```
Comment (shenli) [inline ddl/index.go]: ditto

File: config/config.go [hunk 5]
```diff
@@ -84,6 +84,9 @@ type Log struct {
 	SlowThreshold      uint   `toml:"slow-threshold" json:"slow-threshold"`
 	ExpensiveThreshold uint   `toml:"expensive-threshold" json:"expensive-threshold"`
 	QueryLogMaxLen     uint   `toml:"query-log-max-len" json:"query-log-max-len"`
+
+	// DDLSlowOprThreshold is the threshold of ddl slow operations.
+	DDLSlowOprThreshold uint `toml:"ddl-slow-opr-threshold" json:"ddl-slow-opr-threshold"`
```
Comment (coocood) [inline config/config.go]: I think if we make this configurable, it's better to use HTTP API, it doesn't require a restart of the server.

Comment (coocood) [reply config/config.go]: Then the config option can be removed.

File: ddl/index.go (hunk 4 above)
Comment (zimulala) [reply ddl/index.go]: Could we define a constant for `3000` ？

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [reply ddl/index.go]: Ditto.

**Comment by winkyao** [comment]
@shenli @coocood PTAL

File: ddl/index.go [hunk 6]
```diff
@@ -624,6 +641,7 @@ func (w *addIndexWorker) backfillIndexInTxn(handleRange reorgIndexTask) (nextHan
 		}
 		return nil
 	})
+	w.logSlowOperations(time.Since(oprStartTime), "backfillIndexInTxn", 3000)
```
Comment (shenli) [inline ddl/index.go]: This is a constant threshold?

Comment (winkyao) [reply ddl/index.go]: I think 3s is large enough, so no needs to be configurable.

**Comment by coocood** [comment]
LGTM

**Pull request merged**